### PR TITLE
notification

### PR DESCRIPTION
**Pull request by SudhansuSekharJena**

### CODE REVIEW

File: NotificationModule/settings.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    },
+]
+
+WSGI_APPLICATION = 'NotificationModule.wsgi.application'
+
+
+# Database
+# https://docs.djangoproject.com/en/5.0/ref/settings/#databases
+
+DATABASES = {
```
Comment (mrmjpatra) [inline NotificationModule/settings.py:79]: all the database credentials comes from .env

File: NotifyMe/models/notification.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+from django.db import models
+from django.utils import timezone
+from datetime import timedelta
+from .notificationType import NotificationType
+
+#-------NOTIFICATION MODEL---------
+class Notification(models.Model):
+  title = models.CharField(max_length=40)
```
Comment (mrmjpatra) [inline NotifyMe/models/notification.py:8]: all this values should come from setting.py e.g. declare NOITIFICATION_NAME_LENGTH=60
```suggestion
  title = models.CharField(max_length=40)
```

File: NotifyMe/models/notificationType.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+from django.db import models
+from datetime import timedelta
+
+
+class NotificationType(models.Model):
+  notification_type = models.CharField(max_length = 60)
```
Comment (mrmjpatra) [inline NotifyMe/models/notificationType.py:6]: here length should come from the setting.py file

File: NotifyMe/models/subscription.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+from django.db import models
+from datetime import timedelta
+from django.utils import timezone
+
+from .user import User
+from .subscriptionPlan import SubscriptionPlan
+
+class Subscription(models.Model):
```
Comment (mrmjpatra) [inline NotifyMe/models/subscription.py:8]: create Base model class there declare create at and update_at from that class using inheritance create all the models

File: NotifyMe/models/user.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+from django.db import models
+from datetime import timedelta 
+from django.utils import timezone
+
+from .subscriptionPlan import SubscriptionPlan
+
+class User(models.Model):
```
Comment (mrmjpatra) [inline NotifyMe/models/user.py:7]: use same base class here also

File: NotifyMe/services/service.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+from NotifyMe.models.subscription import Subscription
+from NotifyMe.models.subscriptionPlan import SubscriptionPlan
+from NotifyMe.models.user import User
+
+class UserService:
+  def get_user_data(self, request):
+    try:
+      users = User.objects.all()
+      return users
+    except Exception as e:
```
Comment (mrmjpatra) [inline NotifyMe/services/service.py:12]: handle other exceptiona aks chat gpt , Exception should be the last to catch , use custom exception

File: NotifyMe/services/service.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    try:
+      subscriptions = Subscription.objects.all()
+      return subscriptions
+    except Exception as e:
+      print(f"error:{e}")
+      return None
+      
+class NotificationService:
+  def get_notification_data(self, request):
+    try:
```
Comment (mrmjpatra) [inline NotifyMe/services/service.py:27]: add other exceptions

File: NotifyMe/views/views.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+from NotifyMe.models.subscriptionPlan import SubscriptionPlan
+from rest_framework import status 
+from rest_framework.views import APIView
+from NotifyMe.services.service import *
+
+
+# ---------USER-API---------------- 
+# DONOT CALL MODEL INSIDE VIEW---> ONLY CALL SERVICE AND SERIALIZER
+
+class UserAPI(APIView, UserService):
```
Comment (mrmjpatra) [inline NotifyMe/views/views.py:18]: why are extending from UserService

File: NotifyMe/views/views.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+  
+  def get(self, request):
+    try:
+      objects = self.get_user_data(request)
+      serializer = UserSerializer(objects, many=True)
+      return Response(serializer.data)
+    except Exception as e:
```
Comment (mrmjpatra) [inline NotifyMe/views/views.py:25]: handle other exception

File: NotifyMe/views/views.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+      print(f"Error: {e}")
```
Comment (mrmjpatra) [inline NotifyMe/views/views.py:26]: dont print use logger to store all info and exception related errors

File: NotifyMe/views/views.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+      return Response({"error": str(e)}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
```
Comment (mrmjpatra) [inline NotifyMe/views/views.py:27]: define fixed response type in all the api response refer any article to get response structure

File: NotifyMe/views/views.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+  
+  def post(self, request):
+    try:
+      data = request.data.copy() # dictionary of key and value sent by client.
+      # sid  = data['subscription_plan']
```
Comment (mrmjpatra) [inline NotifyMe/views/views.py:32]: remove commented code

File: NotifyMe/views/views.py [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+      # user = User()
+      # user.email_id = data['email_id']
+      # user.first_name = data['first_name']
+      # user.last_name = data['last_name']  
+      # user.subscription_plan = data['subscription_plan']
+      #--------------------------------------------------
+      
+      try:
+        if 'subscription_plan' in data: # means if client have send any data for subscription field.
+          SubscriptionPlan.objects.get(id=data['subscription_plan'])
```
Comment (mrmjpatra) [inline NotifyMe/views/views.py:45]: dont use model in view use service method

File: NotifyMe/views/views.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+    except User.DoesNotExist:
+      return Response({"error": "User not found"}, status=status.HTTP_404_NOT_FOUND)
+    except Exception as e:
+      print(f"Error: {e}")
+      return Response({"error": str(e)}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+    
+  def delete(self, request):
+    try:
+      data = request.data 
+      obj = User.objects.get(id=data['id'])
```
Comment (mrmjpatra) [inline NotifyMe/views/views.py:94]: dont use model

File: NotifyMe/views/views.py [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+    except Exception as e:
+      print(f"Error: {e}")
+      return Response({"error": str(e)}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+      
+    
+ 
+ 
+#----------SUBSCRIPTION-API----------------      
+    
+class SubscriptionAPI(APIView, SubscriptionService):
```
Comment (mrmjpatra) [inline NotifyMe/views/views.py:108]: dont extend from service

File: NotifyMe/views/views.py [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+#       return Response({"message":"User added successfully"}, status = status.HTTP_201_CREATED)
+    
+#     return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+  
+#   def put(self, request):
+#     data = request.data 
+#     # obj = User.objects.get(id=data['id'])
+#     serializer = UserSerializer(data=data)
+#     if serializer.is_valid():
+#       serializer.save()
```
Comment (mrmjpatra) [inline NotifyMe/views/views.py:225]: remove commented code

File: NotifyMe/serializers.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+from .models.subscription import Subscription
+from .models.subscriptionPlan import SubscriptionPlan
+from .models.user import User
+from django.utils import timezone
+from datetime import datetime
+from datetime import timedelta
+from .models.notificationType import NotificationType
+
+class UserSerializer(serializers.ModelSerializer):
+    subscription_plan = serializers.PrimaryKeyRelatedField(queryset=SubscriptionPlan.objects.all())
```
Comment (mrmjpatra) [inline NotifyMe/serializers.py:14]: dont write complicated code make it user understandable code

File: NotifyMe/serializers.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+        depth=1
+    
+    # THIS IS HELPING IN CREATING SUBSCRIPTION INSTANCE FROM THE INSIDE OF USER... 
+    def create(self, validated_data):
+        subscription_plan = validated_data.pop('subscription_plan')
+        user = User.objects.create(subscription_plan=subscription_plan, **validated_data)
+        
+        start_date = timezone.now()
+        
+        if subscription_plan.subscription_plan == "Basic": # start_Date+30days
```
Comment (mrmjpatra) [inline NotifyMe/serializers.py:28]: create a constant.py there declare all the subscription_plan dont compare with hard coded values

File: api/urls.py [hunk 19]
```diff
@@ -0,0 +1,11 @@
+from NotifyMe.views.views import UserAPI, SubscriptionAPI, NotificationAPI
+
+from django.contrib import admin
+from django.urls import path
+
+# url patterns are list of urls
+urlpatterns = [
+  path("user/", UserAPI.as_view()),
```
Comment (mrmjpatra) [inline api/urls.py:8]: dont append slash at the end

**Comment by SudhansuSekharJena** [comment]
Created a new branch

**Pull request closed without merging**